tests/integration/abi_router_test.py

```diff
@@ -0,0 +1,464 @@
+        ],
+        {
+            DRProp.passed: lambda args: args[1] % 2 == 1,
+            DRProp.lastLog: None,
+        },
+    ),
+    (
+        qa_runner(None),
+        [(False, OnComplete.OptInOC)],
+        qc_runner(None),
+        [
+            (False, OnComplete.ClearStateOC),
+        ],
+        {
+            DRProp.passed: True,
+            DRProp.lastLog: lambda _, actual: actual
+            in (None, DryRunEncoder.hex("optin call")),
+        },
+    ),
+]
+
+
+def make_yacc_runner(method, approval):
+    return MethodRunner(
+        f"yacc_{'approval' if approval else 'clear'}",
+        approval,
+        YACC_TEAL if approval else YACC_CLEAR_TEAL,
+        CONTRACT,
+        method,
+    )
+
+
+# strip out the bare calls for YACC ~ "yetAnotherContractConstructedFromRouter":
+def yaccify(q_cases):
+    return [
+        (make_yacc_runner(method, True), c[1], make_yacc_runner(method, False), *c[3:])
+        for c in q_cases
+        if (method := c[0].method)
+    ]
+
+
+YACC_CASES = yaccify(QUESTIONABLE_CASES)
+
+ALL_CASES = QUESTIONABLE_CASES + YACC_CASES
+
+
+def get_aa_strat(method_runner, abi_args_mod=None) -> ABIMethodCallStrategy:
+    return ABIMethodCallStrategy(
+        method_runner.contract,
+        RandomABIStrategyHalfSized,
+        num_dryruns=NUM_ROUTER_DRYRUNS,
+        abi_args_mod=abi_args_mod,
+    )
+
+
+@pytest.mark.parametrize(
+    "approval_runner, approval_call_types, clear_runner, clear_call_types, invariants",
+    ALL_CASES,
+)
+def test_positive(
+    approval_runner, approval_call_types, clear_runner, clear_call_types, invariants
+):
+    """
+    Test the _positive_ version of a case. In other words, ensure that for the given:
+        * method or bare call
+        * OnComplete value
+        * number of arguments
+    that the app call succeeds according to the provided _invariants_ success definition
+    """
+
+    def run_positive(is_approve, method_runner, call_types, invariants):
+        good_abi_args = get_aa_strat(method_runner)
+
+        algod = get_algod()
+
+        if not call_types:
+            return
+
+        method = method_runner.method
+        sim = Simulation(
+            algod,
+            ExecutionMode.Application,
+            method_runner.teal,
+            invariants,
+            abi_method_signature=good_abi_args.method_signature(method),
+        )
+
+        def msg():
+            return f"""
+    TEST CASE [{method_runner}]({"APPROVAL" if is_approve else "CLEAR"}):
+    test_function={inspect.stack()[2][3]}
+    method={method}
+    is_app_create={is_app_create}
+    on_complete={on_complete!r}"""
+
+        for is_app_create, on_complete in call_types:
+            sim_result = sim.run_and_assert(
+                good_abi_args,
+                txn_params=TxParams.for_app(
+                    is_app_create=is_app_create,
+                    on_complete=on_complete,
+                ),
+                msg=msg(),
+            )
```

Comment:
## STEP 3. Call the Simulations's `run_and_assert()` method
It needs:
1. the `abi_args_strategy` created in previous step (2)
2. `txn_params` which will need the `is_app_create` and `on_complete` parameters which indicate what the dry-run's call context is